### PR TITLE
Add the "gen2" map importer.

### LINKED ISSUE #9335: Legacy isometric map import
Work in progress :construction: at

* https://github.com/pchote/OpenRA/tree/ts-map-importer
* ~~https://github.com/Mailaender/OpenRA/tree/ts-map-import~~

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26652688-legacy-isometric-map-import?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by pchote**
Expands #10368 with integration of the low level components and a collection of style fixes.
Closes https://github.com/OpenRA/OpenRA/issues/9335.

### CODE REVIEW

File: OpenRA.Game/FileFormats/LZOCompression.cs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+ *   http://www.oberhumer.com/opensource/lzo/
+ */
+
+#endregion
+
+using System;
+
+namespace OpenRA.FileFormats
+{
+	public static class LZOCompression
```
Comment (pchote) [inline OpenRA.Game/FileFormats/LZOCompression.cs]: Note: I have chosen this name to be consistent with a future planned PR that renames `Format80` &rarr; `LCWCompression` and `Format40` &rarr; `XORDeltaCompression` (which are the proper names for the algorithms).

File: OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+			map.Options = new MapOptions();
+			map.RequiresMod = modData.Manifest.Mod.Id;
+
+			return map;
+		}
+
+		void ReadTiles(Map map, IniFile file)
+		{
+			var tileset = Game.ModData.DefaultRules.TileSets[map.Tileset];
+			var mapSection = file.GetSection("IsoMapPack5");
```
Comment (Mailaender) [inline OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs]: Looks like this is "Gen" 5 in Westwood library nomenclature.

**Comment by pchote** [comment]
Now depends on #10418.

**Comment by Mailaender** [comment]
Crashes when trying to import http://resource.openra.net/maps/6956 again.

```
mono --debug OpenRA.Utility.exe ts --import-gen2-map /home/matthias/Downloads/karasjok/karasjok.mpr
Error: Utility application crashed. See utility.log for details

Unhandled Exception:
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: startIndex
  at System.ThrowHelper.ThrowArgumentOutOfRangeException (ExceptionArgument argument, ExceptionResource resource) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.2.1/external/referencesource/mscorlib/system/throwhelper.cs:90 
  at System.BitConverter.ToUInt16 (System.Byte[] value, Int32 startIndex) [0x00018] in /home/abuild/rpmbuild/BUILD/mono-4.2.1/external/referencesource/mscorlib/system/bitconverter.cs:291 
  at OpenRA.Mods.TS.UtilityCommands.ImportGen2MapCommand.UnpackLZO (System.Byte[] src, System.Byte[] dest) [0x0000d] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs:188 
  at OpenRA.Mods.TS.UtilityCommands.ImportGen2MapCommand.ReadTiles (OpenRA.Map map, OpenRA.FileFormats.IniFile file) [0x00087] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs:248 
  at OpenRA.Mods.TS.UtilityCommands.ImportGen2MapCommand.Run (OpenRA.ModData modData, System.String[] args) [0x00031] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs:163 
  at OpenRA.Utility.Program.Main (System.String[] args) [0x00185] in /home/matthias/Entwicklung/OpenRA/OpenRA.Utility/Program.cs:85 
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: startIndex
  at System.ThrowHelper.ThrowArgumentOutOfRangeException (ExceptionArgument argument, ExceptionResource resource) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.2.1/external/referencesource/mscorlib/system/throwhelper.cs:90 
  at System.BitConverter.ToUInt16 (System.Byte[] value, Int32 startIndex) [0x00018] in /home/abuild/rpmbuild/BUILD/mono-4.2.1/external/referencesource/mscorlib/system/bitconverter.cs:291 
  at OpenRA.Mods.TS.UtilityCommands.ImportGen2MapCommand.UnpackLZO (System.Byte[] src, System.Byte[] dest) [0x0000d] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs:188 
  at OpenRA.Mods.TS.UtilityCommands.ImportGen2MapCommand.ReadTiles (OpenRA.Map map, OpenRA.FileFormats.IniFile file) [0x00087] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs:248 
  at OpenRA.Mods.TS.UtilityCommands.ImportGen2MapCommand.Run (OpenRA.ModData modData, System.String[] args) [0x00031] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs:163 
  at OpenRA.Utility.Program.Main (System.String[] args) [0x00185] in /home/matthias/Entwicklung/OpenRA/OpenRA.Utility/Program.cs:85
```

**Comment by pchote** [comment]
Fixed.

File: OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+using OpenRA.FileSystem;
+using OpenRA.Mods.Common.FileFormats;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.TS.UtilityCommands
+{
+	class ImportGen2MapCommand : IUtilityCommand
+	{
+		public string Name { get { return "--import-gen2-map"; } }
```
Comment (Mailaender) [inline OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs]: I don't think anyone understands what you mean with "gen2" here.

File: OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+
+					// Should be "tib3"
+					0x93, 0x94, 0x95, 0x96, 0x97, 0x98, 0x99, 0x9A, 0x9B, 0x9C,
+					0x9D, 0x9E, 0x9F, 0xA0, 0xA1, 0xA2, 0xA3, 0xA4, 0xA5, 0xA6 } },
+
+			// Veins
+			{ 0x03, new byte[] { 0x7E } }
+		};
+
+		[Desc("FILENAME", "Convert a Tiberian Sun map to the OpenRA format.")]
```
Comment (Mailaender) [inline OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs]: Red Alert 2 should work as well.

**Comment by Mailaender** [comment]
Works as promised: http://resource.openra.net/maps/8376/ :white_check_mark:

File: OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs (hunk 3 above)
Comment (pchote) [inline OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs]: Ok, then.  Renamed back to `--import-ts-map`.

File: OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs (hunk 4 above)
Comment (pchote) [inline OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs]: Not with all of the hardcoded translation tables above.

File: OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs (hunk 3 above)
Comment (Mailaender) [inline OpenRA.Mods.TS/UtilityCommands/ImportGen2MapCommand.cs]: I went with the same naming convention at https://github.com/OpenRA/OpenRA/pull/10404 as well then.

**Comment by Mailaender** [comment]
Needs a rebase due to https://github.com/OpenRA/OpenRA/pull/10418.

**Comment by pchote** [comment]
Rebased.

**Comment by obrakmann** [comment]
:+1: Works nicely. Needs a rebase though (conflict in the .csproj, I assume).

**Comment by pchote** [comment]
Rebased.

**Comment by Mailaender** [comment]
This should also fix the offset issues the previous versions had as reported by @tomsons26. Offset issues

**Pull request merged**